jujugui/static/gui/src/app/components/terminal/terminal.js

```diff
@@ -0,0 +1,126 @@
+/* Copyright (C) 2017 Canonical Ltd. */
+'use strict';
+
+const React = require('react');
+const PropTypes = require('prop-types');
+const shapeup = require('shapeup');
+const XTerm = require('xterm');
+require('xterm/lib/addons/terminado/terminado');
+
+const Lightbox = require('../lightbox/lightbox');
+
+const SvgIcon = require('../svg-icon/svg-icon');
+
+/** Terminal component used to display the Juju shell. */
+class Terminal extends React.Component {
+
+  constructor(props) {
+    super(props);
+    this.state = {opened: false};
+    this.term = null;
+    this.ws = null;
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    const state = this.state;
+    if (prevState.opened && !state.opened) {
+      this.stopTerm();
+      return;
+    }
+    if (!prevState.opened && state.opened) {
+      this.startTerm();
+    }
+  }
+
+  setOpened(opened, evt) {
+    if (evt) {
+      evt.stopPropagation();
+      evt.preventDefault();
+    }
+    this.setState({opened: opened});
+  }
+
+  startTerm() {
+    const props = this.props;
+    const ws = new WebSocket(`ws://${props.address}/ws/`);
```

Comment:
This will need to be wss

Reply:
For sure, but for now the shell server uses ws.